README.md
<img src="https://github.com/tiagosito/loading_mixin/raw/master/doc/assets/loading_mixin_full.png" width="100%" alt="logo" />
<h2 align="center">
  Want to add Loading to your app?<br/>
  With Loading Mixin you get it in a simple, fast and customizable way
</h2>

<p align="center">
<a href="https://pub.dartlang.org/packages/loading_mixin">
  <img alt="Pub Package" src="https://img.shields.io/pub/v/loading_mixin.svg">
</a>
  <a href="https://www.buymeacoffee.com/tiagosito" target="_blank"><img alt="Buy Me A Coffee" src="https://i.imgur.com/aV6DDA7.png"> </a>
</p>
 
  
## Overview

This package is for you to add loading to your application with just two words, adding this mixin to your class.


#### Contributing
  - [https://github.com/tiagosito/loading_mixin](https://github.com/tiagosito/loading_mixin)

#### For more information, see the complete example
  - [https://pub.dev/packages/loading_mixin/example](https://pub.dev/packages/loading_mixin/example)

- **[Getting Started](#getting-started)**

#### Getting Started


#### In _pubspec.yaml_:
```yaml
dependencies:
  loading_mixin: any
```

### Load type: Automatic, Custom, Manual and Predefined Time
<img src="https://github.com/tiagosito/loading_mixin/blob/master/doc/assets/loading_mixin.gif?raw=true" width="30%" alt="Loading Mixin" />

#### Basic:

#### First: in your State class, add the mixin to the LoadingMixin class
```dart
import 'package:loading_mixin/loading_mixin.dart';

class LoadingMixinTest extends StatefulWidget {
  @override
  _LoadingMixinTestState createState() => _LoadingMixinTestState();
}

//State Class
class _LoadingMixinTestState extends State<LoadingMixinTest> with LoadingMixin {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Loading Mixin'),
      ),
      body: buildBody(),
    );
  }
  .
  .
  .
  ```

#### Second: add the startLoad method by passing your asynchronous function wherever you need
```dart
  RaisedButton(onPressed: () async {
      var result = await this.startAutomaticLoad(context, requestData);

      result != null ? print('$result') : print('No data found!');
    },
    child: Text('Default Load',style: TextStyle(color: Colors.white),
    ),
  )

  //Async/HTTP request Example
  static Future requestData() async {
    await Future.delayed(Duration(milliseconds: 1500));
    return 1844;
  }
}
```

#### Custom: You can also customize your loading by passing the customLaod parameter
```dart

  RaisedButton(onPressed: () async {
    var customLoad = Container(
      height: 50,
      width: 50,
      color: Colors.transparent,
      child: Center(
          child: CircularProgressIndicator(
          valueColor: new AlwaysStoppedAnimation<Color>(Colors.yellow),
          ),
      ),
      );
      var result = await this.startAutomaticLoad(
        context,
        requestData,
        loadingConfig: LoadingConfig(customLoad: customLoad),
        );

      result != null ? print('$result') : print('No data found!');
    },
    child: Text(
        'Async',
        style: TextStyle(color: Colors.white),
    ),
  )

```
#### Manual Load: You can also start and finish the load manually
```dart

  RaisedButton( onPressed: () async {
      //Open Load
      var ovelayEntry = this.startManualLoad(context);

      //Your code
      print('Put your code here!');
      await Future.delayed(Duration(seconds: 2));

      //Close Load
      this.endManualLoad(context, ovelayEntry);
    },
    child: Text(
      'Manual Load',
      style: TextStyle(color: Colors.white),
    ),
  )

```

## Features and bugs

Please send feature requests and bugs at the [issue tracker](https://github.com/tiagosito/loading_mixin/issues).


CHANGELOG.md
# loading_mixin

## [1.1.0]

Brakchanges!

## [0.0.4]

Remove cancel.

## [0.0.3]

Added Automatic, Manual and PredefinedTime methods.

## [0.0.2]

Added formatted code and documentation.

## [0.0.1]

First version of Loading Mixin.


test/loading_mixin_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('adds one to input values', () async {
    var result = await requestData();

    expect(result, 1844);
  });
}

//HTTP request Example
Future requestData() async {
  await Future.delayed(Duration(milliseconds: 1500));
  return 1844;
}


lib/loading_mixin.dart
import 'package:flutter/material.dart';

import 'loadig_configuration.dart';

mixin LoadingMixin {
  ///
  /// Method name: [startAutomaticLoad]
  ///
  /// This is the method for calling your asynchronous function
  ///
  /// Parameters:
  ///
  /// BuildContext: [context], A handle to the location of a widget in the widget tree
  ///
  /// Function: [callback], Your asynchronous function
  ///
  /// Widget: [loadingConfig], Create your own custom loading or make your customization
  ///
  ///
  /// Example:
  /// ```
  /// RaisedButton(
  ///  onPressed: () async {
  ///     var result = await this.startAutomaticLoad(context, requestData);
  ///   },
  ///  child: Text('Fetch'),
  /// )
  ///
  /// //HTTP request Example
  /// static Future requestData() async {
  ///   await Future.delayed(Duration(milliseconds: 1500));
  ///   return 1844;
  /// }
  /// ```
  ///
  Future<LoadingResult> startAutomaticLoad<LoadingResult>(
      BuildContext context, Function callback,
      {LoadingConfig loadingConfig}) async {
    OverlayState overlayState = Overlay.of(context);
    OverlayEntry overlayEntry = OverlayEntry(
      builder: (context) => loadingConfig ?? LoadingConfig(),
    );

    try {
      overlayState.insert(overlayEntry);

      LoadingResult value = await callback();

      overlayEntry.remove();
      return value;
    } catch (e) {
      debugPrint('\n\n **************\n\nERROR: $e\n\n');
      overlayEntry.remove();
      return e;
    }
  }

  ///
  /// Method name: [startManualLoad]
  ///
  /// This is the method for calling when you want to start and close the load manually
  ///
  /// Parameters:
  ///
  /// BuildContext: [context], A handle to the location of a widget in the widget tree
  ///
  /// Widget: [loadingConfig], Create your own custom loading or make your customization
  ///
  /// Example:
  /// ```
  /// RaisedButton(
  ///  onPressed: () async {
  ///     //Start Laod
  ///     var ovelayEntry = this.startManualLoad(context);
  ///
  ///     //Code here
  ///     result = await requestData();
  ///
  ///     //Close Laod
  ///     this.endManualLoad(context, ovelayEntry);
  ///   },
  ///  child: Text('Fetch'),
  /// )
  ///
  /// //HTTP request Example
  /// static Future requestData() async {
  ///   await Future.delayed(Duration(milliseconds: 1500));
  ///   return 1844;
  /// }
  /// ```
  ///
  OverlayEntry startManualLoad<T>(BuildContext context,
      {LoadingConfig loadingConfig}) {
    OverlayState overlayState = Overlay.of(context);
    OverlayEntry overlayEntry = OverlayEntry(
      builder: (context) => loadingConfig ?? LoadingConfig(),
    );

    try {
      overlayState.insert(overlayEntry);
      return overlayEntry;
    } catch (e) {
      debugPrint('\n\n **************\n\nERROR: $e\n\n');
      return e;
    }
  }

  ///
  /// Method name: [endManualLoad]
  ///
  /// This method needs to be called after you call the manual method and process what you want
  ///
  /// Parameters:
  ///
  /// BuildContext: [context], A handle to the location of a widget in the widget tree
  ///
  /// Widget: [overlayEntryToRemove], Overlay that will be added and then removed from the screen
  ///
  /// Example:
  /// ```
  /// RaisedButton(
  ///  onPressed: () async {
  ///     //Start Laod
  ///     var ovelayEntry = this.startManualLoad(context);
  ///
  ///     //Code here
  ///     result = await requestData();
  ///
  ///     //Close Laod
  ///     this.endManualLoad(context, ovelayEntry);
  ///   },
  ///  child: Text('Fetch'),
  /// )
  ///
  /// //HTTP request Example
  /// static Future requestData() async {
  ///   await Future.delayed(Duration(milliseconds: 1500));
  ///   return 1844;
  /// }
  /// ```
  ///
  bool endManualLoad<T>(
      BuildContext context, OverlayEntry overlayEntryToRemove) {
    try {
      overlayEntryToRemove.remove();
      return true;
    } catch (e) {
      debugPrint('\n\n **************\n\nERROR: $e\n\n');
      return e;
    }
  }
}


lib/loadig_configuration.dart
import 'dart:ui';

import 'package:flutter/material.dart';

class LoadingConfig extends StatelessWidget {
  ///
  /// double: [blurSigmaX], blur X axis
  ///
  final double blurSigmaX;

  ///
  /// double: [blurSigmaY], blur Y axis
  ///
  final double blurSigmaY;

  ///
  /// double: [widthLoading], loading widget width
  ///
  final double widthLoading;

  ///
  /// double: [heightLoading], widget height
  ///
  final double heightLoading;

  ///
  /// Color: [backgroundColor], background color
  ///
  final Color backgroundColor;

  ///
  /// Color: [circularProgressIndicatorColor], circularProgressIndicator color
  ///
  final Color circularProgressIndicatorColor;

  ///
  /// Widget: [customLoad], custom widget to show loading
  ///
  final Widget customLoad;

  const LoadingConfig({
    Key key,
    double blurSigmaX,
    double blurSigmaY,
    double widthLoading,
    double heightLoading,
    Color backgroundColor,
    this.circularProgressIndicatorColor,
    this.customLoad,
  })  : this.blurSigmaX = blurSigmaX ?? 0,
        this.blurSigmaY = blurSigmaY ?? 0,
        this.widthLoading = widthLoading ?? 50,
        this.heightLoading = heightLoading ?? 50,
        this.backgroundColor = backgroundColor ?? null,
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return Positioned(
      child: Center(
        child: BackdropFilter(
          filter: ImageFilter.blur(
              sigmaX: this.blurSigmaX, sigmaY: this.blurSigmaY),
          child: Container(
            color: this.backgroundColor ??
                Theme.of(context).primaryColor.withOpacity(0.08),
            child: SizedBox.expand(
              child: Center(
                child: Container(
                  width: this.widthLoading,
                  height: this.heightLoading,
                  child: this.customLoad != null
                      ? this.customLoad
                      : CircularProgressIndicator(
                          valueColor: AlwaysStoppedAnimation<Color>(
                              circularProgressIndicatorColor ??
                                  Theme.of(context).primaryColor),
                        ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:loading_mixin/loadig_configuration.dart';
import 'package:loading_mixin/loading_mixin.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        //primarySwatch: Colors.blue,
        primaryColor: Colors.deepPurple,
        buttonColor: Colors.deepPurple,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: LoadingMixinTest(),
    );
  }
}

class LoadingMixinTest extends StatefulWidget {
  @override
  _LoadingMixinTestState createState() => _LoadingMixinTestState();
}

class _LoadingMixinTestState extends State<LoadingMixinTest> with LoadingMixin {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Loading Mixin'),
      ),
      body: buildBody(),
    );
  }

  buildBody() {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Center(
        child: ListView(
          children: [
            RaisedButton(
              onPressed: () async {
                var result = await this.startAutomaticLoad(context, requestData);

                result != null ? print('$result') : print('No data found!');
              },
              child: Text(
                'Default Load',
                style: TextStyle(color: Colors.white),
              ),
            ),
            RaisedButton(
              onPressed: () async {
                var customLoad = Center(
                  child: Center(
                    child: Container(
                      height: 150,
                      width: 150,
                      child: CircularProgressIndicator(
                        strokeWidth: 25,
                        backgroundColor: Colors.deepPurple,
                        valueColor: new AlwaysStoppedAnimation<Color>(Colors.yellow),
                      ),
                    ),
                  ),
                );
                var result = await this.startAutomaticLoad(
                  context,
                  requestData,
                  loadingConfig: LoadingConfig(customLoad: customLoad),
                );

                result != null ? print('$result') : print('No data found!');
              },
              child: Text(
                'Custom Load',
                style: TextStyle(color: Colors.white),
              ),
            ),
            RaisedButton(
              onPressed: () async {
                //Open Load
                var ovelayEntry = this.startManualLoad(context);

                //Your code
                print('Put your code here!');
                await Future.delayed(Duration(seconds: 2));

                //Close Load
                this.endManualLoad(context, ovelayEntry);
              },
              child: Text(
                'Manual Load',
                style: TextStyle(color: Colors.white),
              ),
            ),
          ],
        ),
      ),
    );
  }

  //HTTP request Example
  static Future requestData() async {
    await Future.delayed(Duration(milliseconds: 1500));
    return 1844;
  }
}
